src/main/java/org/folio/processing/mapping/mapper/reader/record/MarcRecordReader.java

```diff
@@ -1,61 +1,178 @@
   private EntityType entityType;
-  private JsonNode fieldsNode = null;
+  private Record marcRecord;
 
   MarcRecordReader(EntityType entityType) {
     this.entityType = entityType;
   }
 
   @Override
-  public void initialize(DataImportEventPayload eventPayload) throws IOException {
-    if (eventPayload.getContext().containsKey(entityType.value())) {
-      String stringRecord = eventPayload.getContext().get(entityType.value());
-      this.fieldsNode = new ObjectMapper().readTree(stringRecord).at(MARC_FIELDS_POINTER);
-    } else {
-      throw new IllegalArgumentException("Can not initialize MarcRecordReader, no suitable entity type found in event payload");
+  public void initialize(DataImportEventPayload eventPayload) throws RuntimeException {
+    try {
+      if (eventPayload.getContext().containsKey(entityType.value())) {
+        String stringRecord = eventPayload.getContext().get(entityType.value());
+        SourceRecord sourceRecord = new JsonObject(stringRecord).mapTo(SourceRecord.class);
+        if (sourceRecord != null
+          && sourceRecord.getParsedRecord() != null
+          && sourceRecord.getParsedRecord().getContent() != null) {
+          MarcReader reader = buildMarcReader(sourceRecord);
+          if (reader.hasNext()) {
+            this.marcRecord = reader.next();
+          } else {
+            throw new IllegalArgumentException("Can not initialize MarcRecordReader, no suitable marc record found in event payload");
+          }
+        }
+      } else {
+        throw new IllegalArgumentException("Can not initialize MarcRecordReader, no suitable entity type found in event payload");
+      }
+    } catch (Exception e) {
+      LOGGER.error("Can not read marc record from context", e);
+      throw e;
     }
   }
 
   @Override
-  public Value read(String ruleExpression) {
-    String tag = getTag(ruleExpression);
-    String condition = getCondition(ruleExpression);
-    return readByCondition(tag, condition);
+  public Value read(MappingRule ruleExpression) {
+    try {
+      if (ruleExpression.getBooleanFieldAction() != null) {
+        return BooleanValue.of(ruleExpression.getBooleanFieldAction());
+      } else if (ruleExpression.getSubfields().isEmpty() && isNotEmpty(ruleExpression.getValue())) {
+        return readSingleField(ruleExpression);
+      } else if (!ruleExpression.getSubfields().isEmpty() && ruleExpression.getRepeatableFieldAction() != null) {
+        return readRepeatableField(ruleExpression);
+      }
+    } catch (Exception e) {
+      LOGGER.error("Error during reading MappingRule expressions ", e);
+    }
+    return MissingValue.getInstance();
   }
 
-  private String getTag(String ruleExpression) {
-    // Stub implementation
-    return ruleExpression;
+  private Value readSingleField(MappingRule ruleExpression) {
+    String[] expressions = ruleExpression.getValue().split(EXPRESSIONS_DIVIDER);
+    boolean arrayValue = ruleExpression.getPath().endsWith("[]");
```

Comment:
I think it will be better to rename this variable. Smth like "isArray". What do you think?